libp2p/NodeTable.cpp

```diff
@@ -0,0 +1,466 @@
+	m_evictionCheckTimer(m_io)
+{
+	for (unsigned i = 0; i < s_bins; i++)
+	{
+		m_state[i].distance = i;
+		m_state[i].modified = chrono::steady_clock::now() - chrono::seconds(1);
+	}
+	
+	m_socketPtr->connect();
+	doRefreshBuckets(boost::system::error_code());
+}
+	
+NodeTable::~NodeTable()
+{
+	m_evictionCheckTimer.cancel();
+	m_bucketRefreshTimer.cancel();
+	m_socketPtr->disconnect();
+}
+
+void NodeTable::join()
+{
+	doFindNode(m_node.id);
+}
+	
+std::list<NodeId> NodeTable::nodes() const
+{
+	std::list<NodeId> nodes;
+	Guard l(x_nodes);
+	for (auto& i: m_nodes)
+		nodes.push_back(i.second->id);
+	return std::move(nodes);
+}
+
+list<NodeTable::NodeEntry> NodeTable::state() const
+{
+	list<NodeEntry> ret;
+	Guard l(x_state);
+	for (auto s: m_state)
+		for (auto n: s.nodes)
+			ret.push_back(*n.lock());
+	return move(ret);
+}
+
+NodeTable::NodeEntry NodeTable::operator[](NodeId _id)
+{
+	Guard l(x_nodes);
+	return *m_nodes[_id];
+}
+
+void NodeTable::requestNeighbors(NodeEntry const& _node, NodeId _target) const
+{
+	FindNode p(_node.endpoint.udp, _target);
+	p.sign(m_secret);
+	m_socketPtr->send(p);
+}
+
+void NodeTable::doFindNode(NodeId _node, unsigned _round, std::shared_ptr<std::set<std::shared_ptr<NodeEntry>>> _tried)
+{
+	if (!m_socketPtr->isOpen() || _round == s_maxSteps)
+		return;
+	
+	if (_round == s_maxSteps)
+	{
+		clog(NodeTableNote) << "Terminating doFindNode after " << _round << " rounds.";
+		return;
+	}
+	else if(!_round && !_tried)
+		// initialized _tried on first round
+		_tried.reset(new std::set<std::shared_ptr<NodeEntry>>());
+	
+	auto nearest = findNearest(_node);
+	std::list<std::shared_ptr<NodeEntry>> tried;
+	for (unsigned i = 0; i < nearest.size() && tried.size() < s_alpha; i++)
+		if (!_tried->count(nearest[i]))
+		{
+			auto r = nearest[i];
+			tried.push_back(r);
+			FindNode p(r->endpoint.udp, _node);
+			p.sign(m_secret);
+			m_socketPtr->send(p);
+		}
+	
+	if (tried.empty())
+	{
+		clog(NodeTableNote) << "Terminating doFindNode after " << _round << " rounds.";
+		return;
+	}
+		
+	while (!tried.empty())
+	{
+		_tried->insert(tried.front());
+		tried.pop_front();
+	}
+	
+	auto self(shared_from_this());
+	m_evictionCheckTimer.expires_from_now(boost::posix_time::milliseconds(c_reqTimeout.count()));
+	m_evictionCheckTimer.async_wait([this, self, _node, _round, _tried](boost::system::error_code const& _ec)
+	{
+		if (_ec)
+			return;
+		doFindNode(_node, _round + 1, _tried);
+	});
+}
+
+std::vector<std::shared_ptr<NodeTable::NodeEntry>> NodeTable::findNearest(NodeId _target)
+{
+	// send s_alpha FindNode packets to nodes we know, closest to target
+	static unsigned lastBin = s_bins - 1;
+	unsigned head = dist(m_node.id, _target);
+	unsigned tail = head == 0 ? lastBin : (head - 1) % s_bins;
+	
+	std::map<unsigned, std::list<std::shared_ptr<NodeEntry>>> found;
+	unsigned count = 0;
+	
+	// if d is 0, then we roll look forward, if last, we reverse, else, spread from d
+	if (head > 1 && tail != lastBin)
```

Comment:
We will have to agree to disagree here. It's code you write so you always have the last say. I am just telling you what I find more readable :)